### PR TITLE
Enhanced support for pop-up date attribute editing

### PR DESCRIPTION
**Pull request by esreli**
This PR introduces enhanced support for pop-up date attribute editing.

1. Introduces new table cell, `PopupAttributeDatePickerCell`, for use by devices running iOS 14. This cell leverages the new iOS 14 `UIDatePicker`, placing the UI element within the attribute cell rather than displaying a date picker in the keyboard area.
1. Introduces support for _time_ as well as date (iOS 13 & 14)

See related issues:
- [[Support for iOS 14]](https://github.com/ArcGIS/open-source-apps/issues/468)
- [[Editing popup date fields to support date and time]](https://github.com/ArcGIS/open-source-apps/issues/472)

### Screenshots

| iOS 13 | iOS 14 - in-line | iOS 14 - expanded |
| --- | --- | --- |
| Simulator Screen Shot - iPhone 11 Pro Max - 2020-11-25 at 14 03 20 | Simulator Screen Shot - iPhone 11 Pro Max - 2020-11-25 at 14 16 24 | Simulator Screen Shot - iPhone 11 Pro Max - 2020-11-25 at 14 16 27 |

| Dark Mode | Dynamic Text Size | 
| --- | --- | 
| Simulator Screen Shot - iPhone 11 Pro Max - 2020-11-25 at 14 00 50 | Simulator Screen Shot - iPhone 11 Pro Max - 2020-11-25 at 14 01 05 |

| iPad - in-line | iPad - expanded |
| --- | --- |
| Simulator Screen Shot - iPad Pro (9 7-inch) - 2020-11-25 at 14 02 33 | Simulator Screen Shot - iPad Pro (9 7-inch) - 2020-11-25 at 14 02 37 |

### CODE REVIEW

File: data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift [hunk 1]
```diff
@@ -113,7 +113,18 @@ extension RichPopupDetailsViewController /* UITableViewDataSource */ {
                     cell = textFieldCell
                     
                 case (.date, _):
-                    cell = tableView.dequeueReusableCell(withIdentifier: "PopupAttributeDateCell", for: indexPath) as! PopupAttributeDateCell
+                    if #available(iOS 14, *) {
+                        cell = tableView.dequeueReusableCell(
+                            withIdentifier: "PopupAttributeDatePickerCell",
+                            for: indexPath
+                        ) as! PopupAttributeDatePickerCell
```
Comment (mhdostal) [inline data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift:120]: The `as! PopupAttributeDatePickerCell` doesn't seem necessary.

File: data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift [hunk 2]
```diff
@@ -113,7 +113,18 @@ extension RichPopupDetailsViewController /* UITableViewDataSource */ {
...
+                    }
+                    else {
+                        cell = tableView.dequeueReusableCell(
+                            withIdentifier: "PopupAttributeDateCell",
+                            for: indexPath
+                        ) as! PopupAttributeDateCell
```
Comment (mhdostal) [inline data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift:126]: Same here, doesn't seem necessary.

File: data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/Table Cells/Attribute Cells/PopupAttributeDatePickerCell.swift [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    
+    var indexPath = IndexPath(row: 0, section: 0)
+    var delegate: PopupAttributeCellDelegate?
+    
+    private var title: String = ""
+    private var value: AttributeValue = ("", nil)
+    
+    override func setEditing(_ editing: Bool, animated: Bool) {
+        super.setEditing(editing, animated: animated)
+        if editing {
```
Comment (mhdostal) [inline data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/Table Cells/Attribute Cells/PopupAttributeDatePickerCell.swift:33]: Can this whole block just be:

```
datePicker.isHidden = !editing
attributeValueLabel.isHidden = editing
```

**Review by mhdostal — CHANGES_REQUESTED** [review]

File: data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift (hunk 1 above)
Comment (esreli) [reply data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift:120]: If you look at line 88, `cell` must adhere to the `PopupAttributeCell` protocol. As such, if I don't explicitly cast this cell, the type falls back to `UITableViewCell`, which does not adhere to the required protocol. I have to leave this as-is.

File: data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift (hunk 2 above)
Comment (esreli) [reply data-collection/data-collection/View Controllers/Rich Popup View Controllers/Rich Popup View Controller/RichPopupDetailsViewController/RichPopupDetailsViewController+UITableViewDataSource.swift:126]: ditto - https://github.com/Esri/data-collection-ios/pull/267#discussion_r532943999

**Comment by esreli** [comment]
Thank you @mhdostal. I've addressed your review, would you have another look?

**Review by mhdostal — APPROVED** [review]

**Comment by esreli** [comment]
@mhdostal i've force pushed a rebase that resolves some merge conflicts. Merging.

**Pull request merged**